### PR TITLE
Give self argument the correct type when using check_untyped_defs

### PR DESCRIPTION
**Pull request by msullivan**
This is accomplished by updating the helper that constructs a callable type
from a func def to fill in the self argument.

This will make check_untyped_defs a lot more useful.

### CODE REVIEW

File: test-data/unit/check-classes.test [hunk 1]
```diff
@@ -6178,3 +6178,15 @@ class B(A):
...
+# flags: --check-untyped-defs
+class Foo:
+    def foo(self):
+        self.bar()
+        self.baz()  # E: "Foo" has no attribute "baz"
+
+    @classmethod
+    def bar(cls):
+        cls.baz()  # E: "Type[Foo]" has no attribute "baz"
+[builtins fixtures/classmethod.pyi]
```
Comment (ilevkivskyi) [inline test-data/unit/check-classes.test]: I would add also tests for two other reported problems:
* An attribute initialized with `0` in untyped `__init__` has revealed type `int` in another untyped method
* Descriptors also work as expected in untyped methods.

File: mypy/typeops.py [hunk 2]
```diff
@@ -369,3 +370,50 @@ def erase_to_union_or_bound(typ: TypeVarType) -> ProperType:
...
+                                 fallback,
+                                 line=func.line, is_ellipsis_args=True)
+            # Return an Overloaded, because some callers may expect that
+            # an OverloadedFuncDef has an Overloaded type.
+            return Overloaded([dummy])
+
+
+def callable_type(fdef: FuncItem, fallback: Instance,
+                  ret_type: Optional[Type] = None) -> CallableType:
+    # FIXME? somewhat unfortunate duplication with prepare_method_signature in semanal
```
Comment (ilevkivskyi) [inline mypy/typeops.py]: I would make this a TODO (without question mark).

File: mypy/typeops.py [hunk 3]
```diff
@@ -369,3 +370,50 @@ def erase_to_union_or_bound(typ: TypeVarType) -> ProperType:
...
+    if fdef.info and not fdef.is_static:
+        self_type = fill_typevars(fdef.info)  # type: Type
```
Comment (ilevkivskyi) [inline mypy/typeops.py]: Another test idea inspired by `fill_typevars()`:
```python
class C(Generic[T]):
    x: T
    def meth(self):
        self.x + 1  # Should be an error with --check-untyped-defs
```

File: test-data/unit/check-classes.test (hunk 1 above)
Comment (JukkaL) [reply test-data/unit/check-classes.test]: Also, if an attribute is initialized to `None` in `__init__` the inferred type should probably be either `Optional[Any]` or `Any` (the first one may generate many false positives). If an attribute is initialized to `[]`, the inferred type should probably be `List[Any]`, and I think that we shouldn't require a type annotation.

**Comment by The-Compiler** [comment]
As a data point: In [my project](https://www.qutebrowser.org/) I just introduced `check_untyped_defs` a week ago - after the upgrade to 0.740 I now see 469 new errors :laughing: 

Thanks for this, though! I have some work ahead of me now, but it indeed seems to make mypy a lot more useful for codebases which aren't fully annotated yet.

**Pull request merged**